art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -283,49 +316,146 @@ def create_model(
         :return model, loss_fn, optimizer - trained model, loss function used to train the model and optimizer used.
         """
-        import torch  # lgtm [py/repeated-import]
-        import torchvision
-
-        device = self.estimator.device
-        model = torchvision.models.ResNet(torchvision.models.resnet.BasicBlock, [2, 2, 2, 2], num_classes=num_classes)
-        loss_fn = torch.nn.CrossEntropyLoss()
-        optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9, weight_decay=5e-4, nesterov=True)
-        model.to(device)
-        y_train = np.argmax(y_train, axis=1)
-        x_tensor = torch.tensor(x_train, dtype=torch.float32, device=device)  # transform to torch tensor
-        y_tensor = torch.tensor(y_train, dtype=torch.long, device=device)
-        x_test = np.transpose(x_test, [0, 3, 1, 2])
-        y_test = np.argmax(y_test, axis=1)
-        x_tensor_test = torch.tensor(x_test, dtype=torch.float32, device=device)  # transform to torch tensor
-        y_tensor_test = torch.tensor(y_test, dtype=torch.long, device=device)
-
-        dataset_train = torch.utils.data.TensorDataset(x_tensor, y_tensor)  # create your dataset
-        dataloader_train = torch.utils.data.DataLoader(dataset_train, batch_size=batch_size)
-
-        dataset_test = torch.utils.data.TensorDataset(x_tensor_test, y_tensor_test)  # create your dataset
-        dataloader_test = torch.utils.data.DataLoader(dataset_test, batch_size=batch_size)
-
-        for epoch in trange(epochs):
-            running_loss = 0.0
-            total = 0
-            accuracy = 0
-            for _, data in enumerate(dataloader_train, 0):
-                inputs, labels = data
-                optimizer.zero_grad()
-                # forward + backward + optimize
-                outputs = model(inputs)
-                loss = loss_fn(outputs, labels)
-                loss.backward()
-                optimizer.step()
-                _, predicted = torch.max(outputs.data, 1)
-                total += labels.size(0)
-                accuracy += (predicted == labels).sum().item()
-                running_loss += loss.item()
-            if (epoch % 5 == 0) or epoch == (epochs - 1):
-                train_accuracy = 100 * accuracy / total
-                logger.info("Epoch %d train accuracy: %f", epoch, train_accuracy)
-        test_accuracy = self.test_accuracy(model, dataloader_test)
-        logger.info("Final test accuracy: %f", test_accuracy)
+        from art.estimators.classification.pytorch import PyTorchClassifier
+        from art.estimators.classification import TensorFlowV2Classifier
+
+        if isinstance(self.substitute_classifier, PyTorchClassifier):
+            import torch
+            from torch import nn
+            from torch.utils.data import TensorDataset, DataLoader
+            import torchvision
+
+            device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

Comment:
Should we use the same device as for the substitute classifier or ask for the GPU ID as input?

Reply:
I think we can keep the same device for the substitute classifier as otherwise it will involve switching between the GPUs?

Reply:
@monshri Yes, let's use the same device.

Reply:
Sure